### ISSUE TITLE
Unable to distribute simulation batches

### ISSUE THREAD

**Issue by kelvin-olaiya** [AlchemistSimulator/Alchemist#1939]
I had a look at the current state of the `alchemist-grid` module. 
I've problems in trying to distribute batches on a single node cluster (aka, the cluster is my laptop).

What i did was:

1. Start Alchemist in server mode. After reading the [Apache Ignite documentation](https://ignite.apache.org/docs/latest/understanding-configuration), I understand that i need to provide a configuration file with Alchemist `-s` CLI option. This is the configuration file i used:

```xml
<?xml version="1.0" encoding="UTF-8"?>

<beans xmlns="http://www.springframework.org/schema/beans"
       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
       xsi:schemaLocation="
        http://www.springframework.org/schema/beans
        http://www.springframework.org/schema/beans/spring-beans.xsd">

    <bean class="org.apache.ignite.configuration.IgniteConfiguration">
        <property name="workDirectory" value="/home/kelvin/Development/ignite"/>
    </bean>
</beans>
```
2. Launch Alchemist in distributed batch mode, providing the same configuration file to the `-d` cli option.

Trying doing the latter results in errors that seem to have to do with serialization. [Here](https://github.com/AlchemistSimulator/Alchemist/files/10673882/log.txt) you can see the error log.

**Comment by DanySK** [comment]
Apparently an object is not serializable: how does Ignite serialize?

**Comment by DanySK** [comment]
Can you deepen the reason why it does refuse to serialize? It looks like it's failing at serializing Kotlin's lambda expressions - maybe we can program it?

**Comment by kelvin-olaiya** [comment]
Ok, I'll try to debug

**Comment by kelvin-olaiya** [comment]
Most of the serialization exceptions seem to origin from classes whose fields in some ways end up dealing with `org.danilopianini.jirf.FunctionEdge`. I guess the problem may be that it does not implement Serializable.

Other classes with similar issue are:
* `org.danilopianini.jirf.FactoryImpl`
* `it.unibo.alchemist.loader.m2m.Context`
* `org.jgrapht.graph.DefaultDirectedGraph` (although this one extends a class who does implement Serializable) 
* `it.unibo.alchemist.loader.m2m.SimulationModel$fromMap$3`

And then there's another kind of exception regarding which I could not find so many information.

```
java.lang.UnsupportedOperationException: can't get field offset on a hidden class: private final kotlin.jvm.functions.Function1 it.unibo.alchemist.loader.m2m.ObjectFactory$$Lambda$145/0x0000000801176228.arg$1
```

**Comment by DanySK** [comment]
Okay, we must change the way we serialize, then, it's too much stuff to make `Serializable`. Also, it may be the case that Ignite internally does not use the Java serialization (and would be right in doing so).
We should send away only the environment, and the environment should not bring a Java Implicit Reflective Factory within, nor the YAML-to-Alchemist model-to-model stuff.

I think that the "strange thing" you see is a Kotlin default argument value.

**Comment by kelvin-olaiya** [comment]
I tried to avoid sending away the `Loader` by making the field in which it is stored `transient` and store in each `RemoteSimulation` an `InitializedEnvironment`. Most of the previous exceptions are gone, but now there's a problem with the serialization of classes with companion objects. Here's the [error log](https://pastebin.com/8dFpTnnH).

**Comment by DanySK** [comment]
Can we replace the serializer of ignite with something else?

**Comment by kelvin-olaiya** [comment]
Yes, there's a method to set a (i believe custom) `Marshaller` but, it is deprecate:
> **Deprecated.** Since 2.1. Some Ignite features will not work if non-null marshaller is set (IgniteCache.withKeepBinary(), .NET, CPP, ODBC)
Sets marshaller to use within grid.

**Comment by kelvin-olaiya** [comment]
> Okay, we must change the way we serialize, then, it's too much stuff to make `Serializable`. Also, it may be the case that Ignite internally does not use the Java serialization (and would be right in doing so). We should send away only the environment, and the environment should not bring a Java Implicit Reflective Factory within, nor the YAML-to-Alchemist model-to-model stuff.

@DanySK I'm working on the distribution of simulations, still having issues with serialization. I'm avoiding using the default Java serialization by using protocol buffers. The doubt that I now have is: to be able to serialize any `Environment` and/or `Exporter` (contained in an `InitializedEnvironment`) using protobuf, or any other framework, I should define a custom serializer for every specific implementation, right? (It seems to me like it would take a lot of effort, I want to be sure I'm not taking the wrong path).

**Comment by DanySK** [comment]
The grid module is now external. It needs to be rethinked.
